types/ember__component/index.d.ts

```diff
@@ -13,11 +13,11 @@ import CoreView from "@ember/component/-private/core-view";
 import ClassNamesSupport from "@ember/component/-private/class-names-support";
 import ViewMixin from "@ember/component/-private/view-mixin";
 import ActionSupport from "@ember/component/-private/action-support";
+import { TemplateFactory } from 'ember-cli-htmlbars';
 
-// tslint:disable-next-line:strict-export-declare-modifiers
-interface TemplateFactory {
-    __htmlbars_inline_precompile_template_factory: any;
-}
+export function setComponentTemplate<T extends object>(template: TemplateFactory, obj: T): T;
```

Comment:
Hm. https://github.com/ember-cli/ember-cli-htmlbars/blob/c63b4ce3b6309f0f24133b32d4efadc2364574ee/lib/index.d.ts#L8-L10

Would it be safer to make this `ReturnType<hbs>`?

Reply:
I had saw that note and considered that. If we really don't want the `TemplateFactory` interface to be exported, we should add `export {}` to that file to shut off auto-export and force the issue. It makes sense to me that the implementation details are irrelevant and that we don't need to export an interface for the return type of `hbs`, but I guess I'm not 100% convinced once we need that type for a parameter as we do here. @chriskrycho want to weigh in?

Reply:
I don't think we should make that call here on behalf of another package, and essentially rely on a bug in their definition IMO. Since they documented that they didn't want it exposed, I don't think we should rely on it being around as they could fix that problem any day.

Reply:
That particular thing was a choice made a loooong time ago. I think we should (a) use `ReturnType` for now, and (b) upstream a fix to `ember-cli-htmlbars` to make that a "nominalish" type you can't construct yourself, roughly so:

```ts
class TemplateFactory {
  private brand: 'TemplateFactory';
}

export type { TemplateFactory };
```

Then we'll be able to actually import it as a type, as it'll be impossible to construct *other* than with the `hbs` helper.

Reply:
Also yes to `export {}` to shut off auto-export. 😅